### ISSUE TITLE
`prod(::Vector{BigInt})` broken on 1.8 and master

### ISSUE THREAD

**Issue by thofma** [JuliaLang/julia#46665]
It would be great if the following could be fixed for 1.8 (it is working fine on <= 1.7).

```julia
julia> prod(BigInt[0, 0])
ERROR: InexactError: check_top_bit(UInt64, -64)
Stacktrace:
  [1] throw_inexacterror(f::Symbol, #unused#::Type{UInt64}, val::Int64)
    @ Core ./boot.jl:614
  [2] check_top_bit
    @ ./boot.jl:628 [inlined]
  [3] toUInt64
    @ ./boot.jl:739 [inlined]
  [4] UInt64
    @ ./boot.jl:769 [inlined]
  [5] convert
    @ ./number.jl:7 [inlined]
  [6] cconvert
    @ ./essentials.jl:412 [inlined]
  [7] init2!
    @ ./gmp.jl:143 [inlined]
  [8] BigInt
    @ ./gmp.jl:56 [inlined]
  [9] prod(arr::Vector{BigInt})
    @ Base.GMP ./gmp.jl:673
 [10] top-level scope
    @ REPL[1]:1
```
The line
https://github.com/JuliaLang/julia/blob/83658d265d815c15baa199782954598592770230/base/gmp.jl#L672
looks wrong, in particular the `unsafe_load(x.d)`.

CC: @rfourquet

**Comment by rfourquet** [comment]
Thanks for the report, there were indeed two bugs there; `unsafe_load(x.d)` in isolation was fine when `x == 0`, it's then guarranteed to return `0` but the whole line gave then a negative number, and otherwise it should have been `unsafe_load(x.d, abs(x.size))`

**Comment by thofma** [comment]
Fixed in #46667, thanks @rfourquet. (Not sure why github did not automatically close the issue.)
